File: LLM-Translations/GPT-4/codenet/Go/Java/s323794848.java
import java.util.*;

public class s323794848 {
    static class Deque {
        LinkedList<Character> front, back;

        Deque() {
            front = new LinkedList<>();
            back = new LinkedList<>();
        }

        boolean isEmptyFront() {
            return front.isEmpty();
        }

        boolean isEmptyBack() {
            return back.isEmpty();
        }

        void pushBack(char v) {
            back.addLast(v);
        }

        void pushFront(char v) {
            front.addLast(v);
        }

        boolean popBack(Character v) {
            boolean r = true;
            if (!isEmptyBack()) {
                v = back.removeLast();
            } else if (!isEmptyFront()) {
                v = front.removeFirst();
            } else {
                r = false;
            }
            return r;
        }

        boolean popFront(Character v) {
            boolean r = true;
            if (!isEmptyFront()) {
                v = front.removeLast();
            } else if (!isEmptyBack()) {
                v = back.removeFirst();
            } else {
                r = false;
            }
            return r;
        }

        char get(int idx) {
            int tot = front.size() + back.size();
            char v = 0;
            if (idx < tot) {
                if (idx < front.size()) {
                    v = front.get(front.size() - 1 - idx);
                } else {
                    v = back.get(idx - front.size());
                }
            }
            return v;
        }
    }

    static class Operation {
        boolean rev, front;
        char c;

        Operation(boolean rev, boolean front, char c) {
            this.rev = rev;
            this.front = front;
            this.c = c;
        }
    }

    static String reverse(String s) {
        StringBuilder r = new StringBuilder(s);
        return r.reverse().toString();
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        String S = in.next();
        int Q = in.nextInt();
        Operation[] ops = new Operation[Q];
        for (int i = 0; i < Q; i++) {
            int q = in.nextInt();
            if (q == 1) {
                ops[i] = new Operation(true, false, ' ');
            } else {
                int f = in.nextInt();
                char c = in.next().charAt(0);
                ops[i] = new Operation(false, f == 1, c);
            }
        }
        int[] sum = new int[Q + 1];
        for (int i = Q - 1; i >= 1; i--) {
            if (ops[i].rev) {
                sum[i - 1] = sum[i] + 1;
            } else {
                sum[i - 1] = sum[i];
            }
        }
        if (sum[0] % 2 == 1) {
            S = reverse(S);
        }
        Deque deq = new Deque();
        for (int i = 0; i < Q; i++) {
            if (!ops[i].rev) {
                if (ops[i].front) {
                    if (sum[i] % 2 == 1) {
                        deq.pushBack(ops[i].c);
                    } else {
                        deq.pushFront(ops[i].c);
                    }
                } else {
                    if (sum[i] % 2 == 1) {
                        deq.pushFront(ops[i].c);
                    } else {
                        deq.pushBack(ops[i].c);
                    }
                }
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = deq.front.size() - 1; i >= 0; i--) {
            sb.append(deq.front.get(i));
        }
        sb.append(S);
        for (int i = 0; i < deq.back.size(); i++) {
            sb.append(deq.back.get(i));
        }
        System.out.println(sb.toString());
    }
}
// 

File: LLM-Translations/GPT-4/codenet/Go/C/s323794848.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct deque {
    char *front, *back;
    int front_size, back_size;
} deque;

bool isEmptyFront(deque *deq) {
    return deq->front_size == 0;
}

bool isEmptyBack(deque *deq) {
    return deq->back_size == 0;
}

void pushBack(deque *deq, char v) {
    deq->back[deq->back_size++] = v;
}

void pushFront(deque *deq, char v) {
    deq->front[deq->front_size++] = v;
}

bool popBack(deque *deq, char *v) {
    bool r = true;
    if (!isEmptyBack(deq)) {
        *v = deq->back[--deq->back_size];
    } else if (!isEmptyFront(deq)) {
        *v = deq->front[--deq->front_size];
    } else {
        r = false;
    }
    return r;
}

bool popFront(deque *deq, char *v) {
    bool r = true;
    if (!isEmptyFront(deq)) {
        *v = deq->front[--deq->front_size];
    } else if (!isEmptyBack(deq)) {
        *v = deq->back[--deq->back_size];
    } else {
        r = false;
    }
    return r;
}

char get(deque *deq, int idx) {
    int tot = deq->front_size + deq->back_size;
    char v = 0;
    if (idx < tot) {
        if (idx < deq->front_size) {
            v = deq->front[deq->front_size - 1 - idx];
        } else {
            v = deq->back[idx - deq->front_size];
        }
    }
    return v;
}

typedef struct op {
    bool rev, front;
    char c;
} op;

void reverse_str(char *s) {
    int len = strlen(s);
    for (int i = 0; i < len / 2; i++) {
        char tmp = s[i];
        s[i] = s[len - 1 - i];
        s[len - 1 - i] = tmp;
    }
}

int main() {
    char S[100010];
    scanf("%s", S);
    int Q;
    scanf("%d", &Q);
    op ops[200010];
    for (int i = 0; i < Q; i++) {
        int q;
        scanf("%d", &q);
        if (q == 1) {
            ops[i].rev = true;
        } else {
            int f;
            scanf("%d %c", &f, &ops[i].c);
            if (f == 1) {
                ops[i].front = true;
            }
        }
    }
    int sum[200010];
    sum[Q] = 0;
    for (int i = Q - 1; i >= 1; i--) {
        if (ops[i].rev) {
            sum[i - 1] = sum[i] + 1;
        } else {
            sum[i - 1] = sum[i];
        }
    }
    if (sum[0] % 2 == 1) {
        reverse_str(S);
    }
    deque deq;
    deq.front = (char *)malloc(100010 * sizeof(char));
    deq.back = (char *)malloc(100010 * sizeof(char));
    deq.front_size = deq.back_size = 0;
    for (int i = 0; i < Q; i++) {
        if (!ops[i].rev) {
            if (ops[i].front) {
                if (sum[i] % 2 == 1) {
                    pushBack(&deq, ops[i].c);
                } else {
                    pushFront(&deq, ops[i].c);
                }
            } else {
                if (sum[i] % 2 == 1) {
                    pushFront(&deq, ops[i].c);
                } else {
                    pushBack(&deq, ops[i].c);
                }
            }
        }
    }
    for (int i = deq.front_size - 1; i >= 0; i--) {
        printf("%c", deq.front[i]);
    }
    printf("%s", S);
    for (int i = 0; i < deq.back_size; i++) {
        printf("%c", deq.back[i]);
    }
    printf("\n");
    free(deq.front);
    free(deq.back);
    return 0;
}


